Improvements to repair item interface

swofle opened the pull request:
#### Summary
```SUMMARY: Interface "Improvement to repair item interface"```

#### Purpose of change
Fixes #27295 
To reduce the number of keystrokes, when player wants to repair multiple items using the same tool.
Also fixed some oddities with the repair interface.

#### Describe the solution
What started off as a straight-forward change (how difficult can it be to add another popup?), quickly turned into a logic mess as I dig deeper.

Anyway, a screenshot of the changed UI:
![screenshot_2019-01-02_00-11-17](https://user-images.githubusercontent.com/44297063/50574425-7456ca80-0e23-11e9-9193-d9f6b591a952.png)

A screenshot of the current interface for comparison:
![screenshot_2019-01-03_21-57-28](https://user-images.githubusercontent.com/44297063/50641383-b5c1b400-0fa2-11e9-9257-392124a53b02.png)

The main change to the repair workflow is to allow player to continue to use the same tool to repair multiple items, without having to repeat the same keystrokes. (ie. Activate a tailor's kit, choose the repair action) This is achieved by adding the new option "Back to item selection" as suggested in #27295. 

Some oddities were observed and amended during testing:
- The game checks if the repair is valid after the moves have been spent. It feels counter-intuitive that player spends some turn to realise there are insufficient charges in tool, or lack of materials for the repair. 
- There is no feedback if player choose the third option "_Repeat until fully repaired.._" when the selected target item has no damage.
- Also the third option is false advertising. When selected, the game will perform a repair attempt before checking that the target item is fully repaired. Player could select this option continuously to reinforce the item. 
- Using Vehicle wielder, will always trigger one repair attempt before the Repeat menu is shown. In fact, it is possible to reinforce the target item successfully in that one attempt (got lucky) without using any moves and ends the repair activity immediately.

I hope to get some feedback if this is intended behaviour:
- Player is unable to practice on reinforced item. Currently, the repair popup will say _Practicing_ if player chose a reinforced item, but selecting any option leads to this feedback - `Your <item> is already enhanced to its maximum potential.` Not sure if it is intended. I left this as it is for now.

**New workflow:**
1. Player starts the repair activity via tool activation.
2. Game checks tool can be used for repair and proceeds to run the new activity.
Beginning of every repair attempt:
3. Check tool is valid.
4. Query player to select repair target and validate target. This is done repeatedly until a valid target is selected.
5. Query player for the desired repair action. This is the Repeat menu shown in the screenshot above. 
- If target has no damage, and player chooses the _fully repaired_ option. Feedback `Your <item> is already fully repaired.` is provided.
- If player chooses the _Back to item selection_ option, go to step 3.
6. Deduct the moves for one repair attempt.
7. Process the repair attempt: determine result, consume charges, handle repair failure etc. 
8. Proceed to the next repair attempt in step 3. Use the same target if it is still valid. Otherwise query target again. Use the same repair action if applicable, otherwise query again.

#### Additional context
This only applies to tailor repair and welding/soldering repair which belong to the same type of activity. Gun repairs and basic repair kits work via a different mechanism, and are not affected by this.

Pupsi-Mupsi commented:
**I like it!**

- One tiny **suggestion.** It would be nice if you could also print out the item-name (_truncated if needed or in a new line_) see mockup below. 


iuse_actor.cpp (line 2878 f.)

```
const std::string &repair_item_actor::action_description( repair_item_actor::repair_type rt )
{
    static const std::array<std::string, NUM_REPAIR_TYPES> arr = {{
            _( "Nothing" ),
            _( "Repairing" ),
            _( "Refitting" ),
            _( "Reinforcing" ),
            _( "Practicing" )
        }
};
```

- And maybe add some color to:

```
        const std::string title = string_format(
                                      _( "%s\nSuccess chance: %.1f%%\nDamage chance: %.1f%%" ),
                                      repair_item_actor::action_description( action_type ).c_str(),
100.0f * chance.first, 100.0f * chance.second );
```

Mockup:
![grafik](https://user-images.githubusercontent.com/44737997/50574209-f7583080-0de3-11e9-8314-51e09269b3ba.png)

swofle commented:
I pondered further about allowing player to practice on reinforced items, and realised that it would likely cost more trouble than benefits. It would mean that after player has reinforced successfully, the game will return to the repair action popup, instead of the item selection (for the next repair). This translates to one additional keystroke when reinforcing multiple items. 
On the other hand, if player wants to practice tailoring, he can always find a higher tier item to do so.

If it is intended that player can't practice on reinforced items, then the second option should be amended to reflect this. Something like "Repair until reinforced" ?

swofle commented:
@Pupsi-Mupsi Good idea. Updated as suggested. Also renamed the second option as I proposed above. Seems that the popup will resize according to the required width.

![screenshot_2019-01-02_00-11-17](https://user-images.githubusercontent.com/44297063/50574425-7456ca80-0e23-11e9-9193-d9f6b591a952.png)

Pupsi-Mupsi commented:
Great!!! Maybe replace the screenshot in your initial post as well, so nobody gets confused about what the result will look like.

Leland commented on src/activity_handlers.cpp:
```diff
@@ -2067,14 +2068,15 @@ repeat_type repeat_menu( const std::string &title, repeat_type last_selection )
+    rmenu.addentry( REPEAT_FOREVER, true, '2', _( "Repeat until reinforced" ) );
     rmenu.addentry( REPEAT_FULL, true, '3', _( "Repeat until fully repaired, but don't reinforce" ) );
     rmenu.addentry( REPEAT_EVENT, true, '4', _( "Repeat until success/failure/level up" ) );
+    rmenu.addentry( REPEAT_INIT, true, '5', _( "Back to item selection" ) );
```
Recommend the simpler `Cancel`

jbytheway commented on src/activity_handlers.cpp:
```diff
@@ -2067,14 +2068,15 @@ repeat_type repeat_menu( const std::string &title, repeat_type last_selection )
 
     rmenu.addentry( REPEAT_ONCE, true, '1', _( "Repeat once" ) );
-    rmenu.addentry( REPEAT_FOREVER, true, '2', _( "Repeat as long as you can" ) );
+    rmenu.addentry( REPEAT_FOREVER, true, '2', _( "Repeat until reinforced" ) );
```
Is this text appearing for items that are not reinforced (like weapons, which are accurized?).  That may be misleading (don't know if it's a big problem, though).

swofle replied:
Interestingly, this repair_item activity only applies for some tools (all the tailoring related tools, soldering iron and the various welders). AFAIK the term "accurized" only applies to gun repair which is not covered here. Melee weapons still use the "reinforced" description. In fact the current title in the popup window will say "reinforcing", so this term here should be valid.

jbytheway commented on src/activity_handlers.cpp:
```diff
@@ -2187,33 +2184,66 @@ void activity_handlers::repair_item_finish( player_activity *act, player *p )
+        // TODO: Allow setting this in the actor
+        // TODO: Don't use charges_to_use: welder has 50 charges per use, soldering iron has 1
+        if( !used_tool->ammo_sufficient() ) {
+            p->add_msg_if_player( _( "Your %s ran out of charges" ), used_tool->tname().c_str() );
```
```suggestion
            p->add_msg_if_player( _( "Your %s ran out of charges" ), used_tool->tname() );
```
(I realize you didn't change this, only moved it, but boyscouting anyway)

swofle replied:
There are multiple usages of c_str() here with string_format as well. Might as well correct them as well. Will do them in a later commit.

jbytheway commented:
In case you're wondering about the `charge_handling carver` test failure: it's nothing to do with your PR; I looked into it and fixed it on #27418.
